File: zee/src/astr.c
/* Dynamically allocated strings
   Copyright (c) 2001-2004 Sandro Sigala.
   Copyright (c) 2003-2006 Reuben Thomas.
   All rights reserved.

   This file is part of Zee.

   Zee is free software; you can redistribute it and/or modify it under
   the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2, or (at your option) any later
   version.

   Zee is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
   for more details.

   You should have received a copy of the GNU General Public License
   along with Zee; see the file COPYING.  If not, write to the Free
   Software Foundation, Fifth Floor, 51 Franklin Street, Boston, MA
   02111-1301, USA.  */

#include "config.h"

#include <assert.h>
#include <stdint.h>

#include "nonstd.h"
#include "zmalloc.h"
#include "astr.h"
#include "rblist.h"


rblist astr_fread(FILE *fp)
{
  int c;
  rblist as = rblist_empty;

  while ((c = getc(fp)) != EOF)
    as = rblist_append(as, c);
  return as;
}

rblist astr_fgets(FILE *fp)
{
  int c;
  rblist as;

  if (feof(fp))
    return NULL;
  as = rblist_empty;
  while ((c = getc(fp)) != EOF && c != '\n')
    as = rblist_append(as, c);
  return as;
}


#ifdef TEST

#include <stdio.h>
#include <stdlib.h>

// Stub to make zrealloc happy
void die(int exitcode)
{
  exit(exitcode);
}

int main(void)
{
  rblist as1, as2, as3;
  FILE *fp;

  as1 = rblist_from_string("hello world!");
  as3 = rblist_sub(as1, 6, 11);
  assert(!rblist_compare(as3, rblist_from_string("world")));

  as2 = rblist_from_string("The ");
  as2 = rblist_concat(as2, as3);
  assert(!rblist_compare(as2, rblist_from_string("The world")));

  as3 = rblist_sub(as1, 6, 11);
  assert(!rblist_compare(as3, rblist_from_string("world")));

  as1 = rblist_from_string("12345");
  as2 = rblist_sub(as1, 3, rblist_length(as1));
  assert(!rblist_compare(as2, rblist_from_string("45")));

  assert(!rblist_compare(rblist_sub(rblist_from_string("12345"), 0, rblist_length(as1)),
                   rblist_from_string("12345")));

  assert(!rblist_compare(rblist_concat(rblist_afmt("%s * %d = ", "5", 3), rblist_afmt("%d", 15)),
                   rblist_from_string("5 * 3 = 15")));

  assert(rblist_compare(rblist_from_string("a"), rblist_from_string("abc")) == -1);

  assert(rblist_compare(rblist_from_string("abc"), rblist_from_string("a")) == 1);

  assert(fp = fopen(SRCPATH "astr.c", "r"));
  as1 = astr_fgets(fp);
  assert(!rblist_compare(as1, rblist_from_string("/* Dynamically allocated strings")));
  assert(fclose(fp) == 0);

  return EXIT_SUCCESS;
}

#endif // TEST

File: zee/astr.lua
-- Efficient string buffers
--
-- Copyright (c) 2011-2014 Free Software Foundation, Inc.
--
-- This file is part of Zee.
--
-- This program is free software; you can redistribute it and/or modify it
-- under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 3, or (at your option)
-- any later version.
--
-- This program is distributed in the hope that it will be useful, but
-- WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

local Object = require "std.object"

alien.default.memchr:types ("pointer", "pointer", "int", "size_t")

local allocation_chunk_size = 16
AStr = Object {
  -- Primitive methods

  _init = function (self, init, len)
    if type (init) == "string" then
      self.buf = alien.array ("char", #init, alien.buffer (init))
      self.length = #init
    elseif type (init) == "number" then
      self.buf = alien.array ("char", math.max (init, 1))
      self.length = init
    elseif type (init) == "userdata" then
      self.buf = alien.array ("char", len, alien.buffer (init))
      self.length = len
    else
      self.buf = alien.array ("char", init)
      self.length = #init
    end
    return self
  end,

  __tostring = function (self)
    return self.buf.buffer:tostring (#self)
  end,

  __index = function (self, n)
    return self.buf[n]
  end,

  __len = function (self)
    return self.length
  end,

  sub = function (self, from, to)
    to = to or self.length
    return AStr (self:topointer (from), to - from + 1)
  end,

  set_len = function (self, n)
    if n > self.buf.length or n < self.buf.length / 2 then
      self.buf:realloc (n + allocation_chunk_size)
    end
    self.length = n
  end,

  move = function (self, to, from, n)
    assert (math.max (from, to) + n <= #self + 1)
    alien.memmove (self:topointer (to), self:topointer (from), n)
  end,

  set = function (self, from, c, n)
    assert (from + n <= #self + 1)
    alien.memset (self:topointer (from), c:byte (), n)
  end,

  remove = function (self, from, n)
    local b, e, eob = from, from + n, #self + 1
    assert (e <= eob)
    self:move (e, b, eob - e)
    self:set_len (#self - n)
  end,

  insert = function (self, from, n)
    assert (from <= #self + 1)
    self:set_len (#self + n)
    self:move (from + n, from, #self + 1 - (from + n))
    self:set (from, '\0', n)
  end,

  replace = function (self, from, rep, len)
    local len = len or #rep
    assert (from + len <= #self + 1)
    if type (rep) ~= "userdata" and type (rep) ~= "string" then
      rep = rep:topointer ()
    end
    alien.memmove (self:topointer (from), rep, len)
    return self
  end,

  rchr = function (self, c, from)
    local b = self.buf
    for i = from - 1, 1, -1 do
      if b[i] == c then return i end
    end
  end,

  start_of_line = function (self, o)
    local prev = self:rchr (string.byte ('\n'), o)
    return prev and prev + 1 or 1
  end,

  end_of_line = function (self, o)
    local next = alien.default.memchr (self:topointer (o), string.byte ('\n'), #self - (o - 1))
    return next and self.buf.buffer:tooffset (next) or #self + 1
  end,


  -- Derived methods

  topointer = function (self, offset)
    return self.buf.buffer:topointer (offset)
  end,

  cat = function (self, src)
    local oldlen = #self
    self:insert (oldlen + 1, #src)
    return self:replace (oldlen + 1, src:topointer (), #src)
  end,

  prev_line = function (self, o)
    local so = self:start_of_line (o)
    return so ~= 1 and self:start_of_line (so - 1) or nil
  end,

  next_line = function (self, o)
    local eo = self:end_of_line (o)
    return eo <= #self and eo + 1 or nil
  end,

  lines = function (self)
    local lines = -1
    local next = 1
    repeat
      next = self:next_line (next)
      lines = lines + 1
    until not next
    return lines
  end,
}

local have_memrchr = pcall (loadstring 'alien.default.memrchr:types ("pointer", "pointer", "int", "size_t")')
if have_memrchr then
  AStr.rchr = function (self, c, from)
    local p = alien.default.memrchr(self:topointer (), c, from - 1)
    return p and self.buf.buffer:tooffset (p)
  end
end
